lib/matplotlib/figure.py

```diff
@@ -1346,27 +1346,44 @@ def add_subplot(self, *args, **kwargs):
             ax1.remove()  # delete ax1 from the figure
             fig.add_subplot(ax1)  # add ax1 back to the figure
         """
-        if not len(args):
-            args = (1, 1, 1)
-
-        if len(args) == 1 and isinstance(args[0], Integral):
-            if not 100 <= args[0] <= 999:
-                raise ValueError("Integer subplot specification must be a "
-                                 "three-digit number, not {}".format(args[0]))
-            args = tuple(map(int, str(args[0])))
-
         if 'figure' in kwargs:
             # Axes itself allows for a 'figure' kwarg, but since we want to
             # bind the created Axes to self, it is not allowed here.
             raise TypeError(
                 "add_subplot() got an unexpected keyword argument 'figure'")
 
-        if isinstance(args[0], SubplotBase):
+        nargs = len(args)
+        if nargs == 0:
+            args = (1, 1, 1)
+        elif nargs == 1:
+            print(type(args[0]))
+            if isinstance(args[0], Integral):
+                if not 100 <= args[0] <= 999:
+                    raise ValueError(f"Integer subplot specification must be "
+                                     f"a three-digit number, not {args[0]}")
+                args = tuple(map(int, str(args[0])))
+            elif isinstance(args[0], (SubplotBase, SubplotSpec)):
+                pass  # no further validation or normalization needed
+            else:
+                raise TypeError('Positional arguments are not a valid '
+                                'position specification.')
+        elif nargs == 3:
+            for arg in args:
+                if not isinstance(arg, Integral):
+                    cbook.warn_deprecated(
+                        "3.3",
+                        message="Passing non-integers to as three-element "
+                                "position specification is deprecated.")
+            args = tuple(map(int, args))
+        else:
+            raise TypeError('Positional arguments are not a valid position '
+                            'specification.')
 
+        if isinstance(args[0], SubplotBase):
```

Comment:
This type is not documented; should it be?

Reply:
It's not documented as type, but in the description:

~~~
            In rare circumstances, `.add_subplot` may be called with a single
            argument, a subplot axes instance already created in the
            present figure but not in the figure's list of axes.
~~~

I'm not even sure what this is good for or how to achieve that, but that code path triggers on the tests. I've intentionally omitted this form the types. I assume adding it to the types without big warnings would probably do more harm than it's worth, because people will start to put axes from other figures in.